x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/OrdinalsGroupingOperator.java

```diff
@@ -375,16 +377,27 @@ void addInput(IntVector docs, Page page) {
                     currentReader = new BlockOrdinalsReader(docValuesSupplier.get(), blockFactory);
                 }
                 try (IntBlock ordinals = currentReader.readOrdinalsAdded1(docs)) {
-                    for (int p = 0; p < ordinals.getPositionCount(); p++) {
-                        int start = ordinals.getFirstValueIndex(p);
-                        int end = start + ordinals.getValueCount(p);
-                        for (int i = start; i < end; i++) {
-                            long ord = ordinals.getInt(i);
+                    IntVector ordinalsVector = ordinals.asVector();
+                    if (ordinalsVector != null) {
+                        for (int p = 0; p < ordinalsVector.getPositionCount(); p++) {
+                            long ord = ordinalsVector.getInt(p);
                             visitedOrds.set(ord);
                         }
-                    }
-                    for (GroupingAggregatorFunction.AddInput addInput : prepared) {
-                        addInput.add(0, ordinals);
+                        for (GroupingAggregatorFunction.AddInput addInput : prepared) {
```

Comment:
I tend to move each `for` loop into a method. I think it's usually more readable that way. Meh.

Reply:
++ I made two methods in https://github.com/elastic/elasticsearch/pull/108118/commits/2e30b3cb6638b7faeebe540b54c774bcfcc79987